File: 1-arrays/cpp/66-plus-one.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> plusOne(vector<int>& digits) {
    int curDigit = digits.size()-1;

    while(digits[curDigit] == 9){
        if(curDigit == 0){
            digits[0] = 0;
            digits.push_back(0);
        } else{
            digits[curDigit] = 0;
            curDigit--;
        }
    }
    digits[curDigit]++;
    return digits;
}

int main(){
    vector<int> nums = {4,3,2,1};
    vector<int> ans = plusOne(nums);

    for(auto i: ans){
        cout<< i<< " ";
    }
    cout << endl;
}

File: 1-arrays/java/66-plus-one.java
import java.util.Arrays;
public class Solution{
    public static int[] plusOne(int[] digits) {
        int curDigit = digits.length-1;

        while(digits[curDigit] == 9){
            if(curDigit == 0){
                int[] newArray = new int[digits.length+1];
                newArray[0] = 1;
                return newArray;
            }
            digits[curDigit] = 0;
            curDigit--;
        }
        digits[curDigit]++;
        return digits;
    }
    public static void main(String args[]){
        int nums[] = {4,3,2,9};
        System.out.println(Arrays.toString(plusOne(nums)));
    }
}
// Time Complexity: O(n)
// Space Complexity: O(n)